pkg/services/sqlstore/permissions/dashboard.go

```diff
@@ -148,15 +148,15 @@ func (f *accessControlDashboardPermissionFilter) buildClauses() {
 
 		if len(toCheck) > 0 {
 			if !useSelfContainedPermissions {
-				builder.WriteString("(dashboard.uid IN (SELECT substr(scope, 16) FROM permission WHERE scope LIKE 'dashboards:uid:%'")
+				builder.WriteString("(dashboard.uid IN (SELECT identifier FROM permission WHERE kind = 'dashboards'")
 				builder.WriteString(rolesFilter)
 				args = append(args, params...)
 
 				if len(toCheck) == 1 {
 					builder.WriteString(" AND action = ?")
 					args = append(args, toCheck[0])
 				} else {
-					builder.WriteString(" AND action IN (?" + strings.Repeat(", ?", len(toCheck)-1) + ") GROUP BY role_id, scope HAVING COUNT(action) = ?")
+					builder.WriteString(" AND action IN (?" + strings.Repeat(", ?", len(toCheck)-1) + ") GROUP BY role_id, scope, identifier HAVING COUNT(action) = ?")
```

Comment:
[question] I may be missing some context here, but why do we add `identifier` to `GROUP BY`? I don't see we were grouping by the substring previously.

If I interpret correctly, the query was meant to return a row per granted scope to the role, whereas adding `identifier` would return a row per granted scope to the role for each object on which it has the scope granted on, regardless of the identifying attribute of the target object or the kind of object.

Reply:
That's an awkward one. In MySQL and SQLite things just seem to work and we don't need `identifier` in `GROUP BY`. Postgres, however, fails with an error like:
> column "permission.identifier" must appear in the GROUP BY clause or be used in an aggregate function
Adding it to the GROUP BY helps. Quick googling/stackoverflowing explains it's the right way and "it works in MySQL because MySQL is broken".

Reply:
I think, what's happening here is: we get all the UIDs from `permission` table where `role` fits any of the user/team roles AND `action` is in the list of the allowed actions. We then check if the requested dashboard/folder UID is in the list of the results from the forementioned query. We usually get one UID per permission but for multiple actions we might have duplicates. Using GROUP BY here eliminates them and leaves dashboard/folder UIDs being unique.

We might also use SELECT DISTINCT ON, but GROUP BY seems cleaner and less dialect-specific.